src/txHashBuilder.c

```diff
@@ -133,34 +169,139 @@ void txHashBuilder_enterOutputs(tx_hash_builder_t* builder)
 	builder->state = TX_HASH_BUILDER_IN_OUTPUTS;
 }
 
-void txHashBuilder_addOutput(
+void txHashBuilder_addOutput_basicData(
         tx_hash_builder_t* builder,
         const uint8_t* addressBuffer, size_t addressSize,
-        uint64_t amount
+        uint64_t amount,
+        uint16_t numAssetGroups
 )
 {
+	_TRACE("state = %u, remainingOutputs = %u", builder->state, builder->remainingOutputs);
+
 	ASSERT(builder->state == TX_HASH_BUILDER_IN_OUTPUTS);
 	ASSERT(builder->remainingOutputs > 0);
 	builder->remainingOutputs--;
 
-	// Array(2)[
-	//   Bytes[address]
-	//   Unsigned[amount]
+	if (numAssetGroups == 0) {
+		// Array(2)[
+		//   Bytes[address]
+		//   Unsigned[amount]
+		// ]
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+			{
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, addressSize);
+				BUILDER_APPEND_DATA(addressBuffer, addressSize);
+			}
+			{
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, amount);
+			}
+		}
+		builder->state = TX_HASH_BUILDER_IN_OUTPUTS;
+	} else {
+		builder->outputData.remainingAssetGroups = numAssetGroups;
+		// Array(2)[
+		//   Bytes[address]
+		//   Array(2)[]
+		//     Unsigned[amount]
+		//     Map(numAssetGroups)[
+		//       // entries added later, { * policy_id => { * asset_name => uint } }
+		//     ]
+		//   ]
+		// ]
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+			{
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, addressSize);
+				BUILDER_APPEND_DATA(addressBuffer, addressSize);
+			}
+			{
```

Comment:
conceptually, shouldn't the `BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);` be besides the previous two APPEND statements, hence within the same scope?

Reply:
Each of the two elements of the top-level array has its own scope. The scope is not used to encompass all array elements (that's the role of the outermost scope).